### PR TITLE
New code example for the "var pattern" (issue #12736)

### PR DESCRIPTION
**Pull request by mssteele**
Addresses #12736 

Documentation changes relating to the new code example.

Also see my separate change to the code sample itself in the dotnet/samples repo.
PR: https://github.com/dotnet/samples/pull/2019

### CODE REVIEW

File: docs/csharp/language-reference/keywords/is.md [hunk 1]
```diff
@@ -101,16 +101,22 @@ The following example shows a comparison of `null` checks:
...
 
-The `var` pattern is a catch-all for any type or value. The value of *expr* is always assigned to a local variable the same type as the compile time type of *expr*. The result of the `is` expression is always `true`. Its syntax is:
+A pattern match with the `var` pattern always succeeds. Its syntax is:
 
 ```csharp
    expr is var varname
 ```
 
-The following example uses the var pattern to assign an expression to a variable named `obj`. It then displays the value and the type of `obj`.
+where the value of *expr* is always assigned to a local variable named *varname*. *varname* is a variable of the same type as the compile time type of *expr*.
```
Comment (pkulikov) [inline docs/csharp/language-reference/keywords/is.md]:
```suggestion
where the value of *expr* is always assigned to a local variable named *varname*. *varname* is a variable of the same type as the compile-time type of *expr*. 
```

File: docs/csharp/language-reference/keywords/is.md [hunk 2]
```diff
@@ -101,16 +101,22 @@ The following example shows a comparison of `null` checks:
...
+A pattern match with the `var` pattern always succeeds. Its syntax is:
 
 ```csharp
    expr is var varname
 ```
 
-The following example uses the var pattern to assign an expression to a variable named `obj`. It then displays the value and the type of `obj`.
+where the value of *expr* is always assigned to a local variable named *varname*. *varname* is a variable of the same type as the compile time type of *expr*. 
+
+If *expr* is `null`, the `is` expression is still true and assigns `null` to *varname*.
```
Comment (pkulikov) [inline docs/csharp/language-reference/keywords/is.md]:
```suggestion
If *expr* evaluates to `null`, the `is` expression produces `true` and assigns `null` to *varname*. 
```

File: docs/csharp/language-reference/keywords/is.md [hunk 3]
```diff
@@ -101,16 +101,22 @@ The following example shows a comparison of `null` checks:
...
 ```csharp
    expr is var varname
 ```
 
-The following example uses the var pattern to assign an expression to a variable named `obj`. It then displays the value and the type of `obj`.
+where the value of *expr* is always assigned to a local variable named *varname*. *varname* is a variable of the same type as the compile time type of *expr*. 
+
+If *expr* is `null`, the `is` expression is still true and assigns `null` to *varname*. 
+
+One reason to use the `var` pattern is to create a temporary variable within a Boolean expression. For example:
```
Comment (pkulikov) [inline docs/csharp/language-reference/keywords/is.md]:
```suggestion
You can use the `var` pattern to create a temporary variable within a Boolean expression, as the following example shows:
```

File: docs/csharp/language-reference/keywords/is.md [hunk 4]
```diff
@@ -101,16 +101,22 @@ The following example shows a comparison of `null` checks:
...
 
 [!code-csharp[is#8](../../../../samples/snippets/csharp/language-reference/keywords/is/is-var-pattern8.cs#8)]
 
+Here the temporary variable is used to store the result of an expensive operation. The variable can then be used multiple times.
```
Comment (pkulikov) [inline docs/csharp/language-reference/keywords/is.md]:
```suggestion
In the preceding example, the temporary variable is used to store the result of an expensive operation. The variable can then be used multiple times.
```

**Review by pkulikov — COMMENTED** [review]
Rephrased to match the usual docs language.

File: docs/csharp/language-reference/keywords/is.md [hunk 5]
```diff
@@ -101,16 +101,22 @@ The following example shows a comparison of `null` checks:
...
+A pattern match with the `var` pattern always succeeds. Its syntax is:
 
 ```csharp
    expr is var varname
 ```
 
-The following example uses the var pattern to assign an expression to a variable named `obj`. It then displays the value and the type of `obj`.
+where the value of *expr* is always assigned to a local variable named *varname*. *varname* is a variable of the same type as the compile-time type of *expr*. 
+
+If *expr* evaluates to `null`, the `is` expression produces `true` and assigns `null` to *varname*.
```
Comment (BillWagner) [inline docs/csharp/language-reference/keywords/is.md]: Is it worth adding a clarifying comment here?  In most situations `is` produces `false` for `null`:

```suggestion
If *expr* evaluates to `null`, the `is` expression produces `true` and assigns `null` to *varname*.  The var pattern is one of the few uses of `is` that produces `true` for a `null` value.
```

**Review by BillWagner — CHANGES_REQUESTED** [review]
I really like this change @mssteele 

I had one suggestion, because we've seen confusion because `is` produces `false` in most cases when the argument is `null`. What do you think?

Marking as "request changes" so it stays in my queue while we discuss.

File: docs/csharp/language-reference/keywords/is.md (hunk 5 above)
Comment (Thraka) [reply docs/csharp/language-reference/keywords/is.md]: I don't think it's particularly useful. Side question, are there other uses of var that does produce true for null?

Comment (BillWagner) [reply docs/csharp/language-reference/keywords/is.md]: @Thraka Nope, `thing is var newVariableName` always succeeds, and always returns true.

But, `thing is string message` returns `false` if `thing` is `null`.

**Comment by mssteele** [comment]
Yes good suggestion @BillWagner 
I think it is worth pointing out. It is a bit unintuitive when you compare it to your example where
`thing is string message` returns `false` if `thing` is `null` 
(Sorry - I should have probably commented next to your suggestion but I accepted it so then couldn't comment there!)

**Review by BillWagner — APPROVED** [review]

**Comment by BillWagner** [comment]
Thanks for making these additional changes @mssteele 

I've approved this, and I'll :shipit: now.

Thanks again!

**Pull request merged**